File: simulations-java/simulations/acid-base-solutions/src/edu/colorado/phet/acidbasesolutions/view/BeakerNode.java
/* Copyright 2010, University of Colorado */

package edu.colorado.phet.acidbasesolutions.view;

import java.awt.*;
import java.awt.geom.GeneralPath;
import java.awt.geom.Line2D;
import java.awt.geom.Rectangle2D;
import java.text.MessageFormat;

import edu.colorado.phet.acidbasesolutions.constants.ABSStrings;
import edu.colorado.phet.acidbasesolutions.model.ABSModel;
import edu.colorado.phet.acidbasesolutions.model.Beaker;
import edu.colorado.phet.common.phetcommon.view.util.PhetFont;
import edu.umd.cs.piccolo.nodes.PPath;
import edu.umd.cs.piccolo.nodes.PText;
import edu.umd.cs.piccolox.nodes.PComposite;

/**
 * Visual representation of a beaker that is filled to the top with a solution.
 * Origin is at the bottom center.
 *
 * @author Chris Malley (cmalley@pixelzoom.com)
 */
public class BeakerNode extends PComposite {
    
    private static final double MAX_VOLUME = 1; // L
    
    private static final Stroke STROKE = new BasicStroke( 6f, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND );
    private static final Color STROKE_COLOR = Color.BLACK;
    
    private static final String[] MAJOR_TICK_LABELS = { null, "1" }; // nothing, 1L
    private static final Color TICK_COLOR = Color.BLACK;
    private static final double MINOR_TICK_SPACING = 0.1; // L
    private static final int MINOR_TICKS_PER_MAJOR_TICK = 5;
    private static final double MAJOR_TICK_LENGTH = 30;
    private static final double MINOR_TICK_LENGTH = 15;
    private static final Stroke MAJOR_TICK_STROKE = new BasicStroke( 2f, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL );
    private static final Stroke MINOR_TICK_STROKE = new BasicStroke( 2f, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL );
    private static final Font TICK_LABEL_FONT = new PhetFont( 24 );
    private static final double TICK_LABEL_X_SPACING = 8;
    private static final double RIM_OFFSET = 20;
    
    private final Beaker beaker;
    private final PPath outlineNode, solutionNode;
    private final GeneralPath outlinePath; 
    private final Rectangle2D solutionRectangle;
    private final PComposite ticksNode;
    
    public BeakerNode( ABSModel model ) {
        super();
        
        this.beaker = model.getBeaker();

        solutionRectangle = new Rectangle2D.Double();
        solutionNode = new PPath();
        solutionNode.setPaint( model.getSolution().getColor() );
        solutionNode.setStroke( null );
        addChild( solutionNode );
        
        outlinePath = new GeneralPath();
        outlineNode = new PPath();
        outlineNode.setPaint( null );
        outlineNode.setStroke( STROKE );
        outlineNode.setStrokePaint( STROKE_COLOR );
        addChild( outlineNode );
        
        ticksNode = new PComposite();
        addChild( ticksNode );
        
        setOffset( beaker.getLocationReference() );
        setVisible( beaker.isVisible() );
        update();
    }
    
    private void update() {
        updateBeaker();
        updateTicks();
        updateSolution();
    }
    
    /*
     * Where the beaker's origin is in relation to its width.
     */
    private double getOriginXOffset() {
        return beaker.getWidth() / 2; // origin a midpoint
    }
    
    /*
     * Where the beaker's origin is in relation to its height.
     */
    private double getOriginYOffset() {
        return beaker.getHeight(); // origin at bottom
    }
    
    /*
     * Creates the shape for the beaker.
     * Origin is at the bottom center.
     * Start drawing the shape at the upper-left corner.
     */
    private void updateBeaker() {
        outlinePath.reset();
        double xOffset = getOriginXOffset();
        double yOffset = getOriginYOffset();
        outlinePath.moveTo( (float) ( -xOffset - RIM_OFFSET ), (float) ( -yOffset - RIM_OFFSET ) );
        outlinePath.lineTo( (float) -xOffset, (float) -yOffset );
        outlinePath.lineTo( (float) -xOffset, (float)( beaker.getHeight() - yOffset ) );
        outlinePath.lineTo( (float) ( beaker.getWidth() - xOffset ), (float)( beaker.getHeight() - yOffset )  );
        outlinePath.lineTo( (float) ( beaker.getWidth() - xOffset ), (float) -yOffset );
        outlinePath.lineTo( (float) ( beaker.getWidth() - xOffset + RIM_OFFSET ), (float) ( -yOffset - RIM_OFFSET ) );
        outlineNode.setPathTo( outlinePath );
    }
    
    private void updateSolution() {
        double xOffset = getOriginXOffset();
        double yOffset = getOriginYOffset();
        solutionRectangle.setRect( -xOffset, -yOffset, beaker.getWidth(), beaker.getHeight() );
        solutionNode.setPathTo( solutionRectangle );
    }
    
    private void updateTicks() {
        ticksNode.removeAllChildren();
        int numberOfTicks = (int) Math.round( MAX_VOLUME / MINOR_TICK_SPACING );
        final double rightX = getOriginXOffset(); // don't use bounds or position will be off because of stroke width
        final double bottomY = beaker.getHeight() - getOriginYOffset(); // don't use bounds or position will be off because of stroke width
        double deltaY = beaker.getHeight() / numberOfTicks;
        for ( int i = 1; i <= numberOfTicks; i++ ) {
            final double y = bottomY - ( i * deltaY );
            if ( i % MINOR_TICKS_PER_MAJOR_TICK == 0 ) {
                // major tick
                Shape tickPath = new Line2D.Double( rightX - MAJOR_TICK_LENGTH, y, rightX - 2, y );
                PPath tickNode = new PPath( tickPath );
                tickNode.setStroke( MAJOR_TICK_STROKE );
                tickNode.setStrokePaint( TICK_COLOR );
                ticksNode.addChild( tickNode );
                
                int labelIndex = ( i / MINOR_TICKS_PER_MAJOR_TICK ) - 1;
                if ( labelIndex < MAJOR_TICK_LABELS.length && MAJOR_TICK_LABELS[ labelIndex ] != null ) {
                    String label = MessageFormat.format( ABSStrings.PATTERN_VALUE_UNITS, MAJOR_TICK_LABELS[ labelIndex ], ABSStrings.LITERS );
                    PText textNode = new PText( label );
                    textNode.setFont( TICK_LABEL_FONT );
                    textNode.setTextPaint( TICK_COLOR );
                    ticksNode.addChild( textNode );
                    double xOffset = tickNode.getFullBounds().getMinX() - textNode.getFullBoundsReference().getWidth() - TICK_LABEL_X_SPACING;
                    double yOffset = tickNode.getFullBounds().getMinY() - ( textNode.getFullBoundsReference().getHeight() / 2 );
                    textNode.setOffset( xOffset, yOffset );
                }
            }
            else {
                // minor tick
                Shape tickPath = new Line2D.Double( rightX - MINOR_TICK_LENGTH, y, rightX - 2, y );
                PPath tickNode = new PPath( tickPath );
                tickNode.setStroke( MINOR_TICK_STROKE );
                tickNode.setStrokePaint( TICK_COLOR );
                ticksNode.addChild( tickNode );
            }
        }
    }
}

File: trunk/simulations-html/simulations/beers-law-lab/js/concentration/view/BeakerNode.js
// Copyright 2013, University of Colorado

define( [
            'easel',
            'common/model/Inheritance',
            'common/util/StringUtils',
            'i18n!../../../nls/beers-law-lab-strings'
        ],
        function ( Easel, Inheritance, StringUtils, Strings ) {

            /**
             * Constructor
             * @param {Beaker} beaker
             * @param {ModelViewTransform2D} mvt
             * @constructor
             */
            function BeakerNode( beaker, mvt ) {

                Easel.Container.call( this ); // constructor stealing

                // constants
                var MAX_VOLUME = 1;
                var RIM_OFFSET = 20;
                var MINOR_TICK_SPACING = 0.1; // L
                var MINOR_TICKS_PER_MAJOR_TICK = 5;
                var MAJOR_TICK_LENGTH = 30;
                var MINOR_TICK_LENGTH = 15;
                var TICK_LABEL_X_SPACING = 8;
                var MAJOR_TICK_LABELS = new Array( "\u00bd", "1" ); // 1/2, 1

                // outline of the beaker, starting from upper left
                var width = mvt.modelToView( beaker.size.width );
                var height = mvt.modelToView( beaker.size.height );
                var outlineNode = new Easel.Shape();
                outlineNode.graphics
                        .setStrokeStyle( 6, 'round', 'round' )
                        .beginStroke( 'black' )
                        .moveTo( -(width / 2 ) - RIM_OFFSET, -height - RIM_OFFSET )
                        .lineTo( -(width / 2), -height )
                        .lineTo( -(width / 2), 0 )
                        .lineTo( width / 2, 0 )
                        .lineTo( width / 2, -height )
                        .lineTo( (width / 2) + RIM_OFFSET, -height - RIM_OFFSET );
                this.addChild( outlineNode );

                // horizontal tick marks, left edge, from bottom up
                var ticksParent = new Easel.Container();
                this.addChild( ticksParent );
                var numberOfTicks = Math.round( MAX_VOLUME / MINOR_TICK_SPACING );
                var deltaY = height / numberOfTicks;
                for ( var i = 1; i <= numberOfTicks; i++ ) {

                    // tick
                    var isMajorTick = ( i % MINOR_TICKS_PER_MAJOR_TICK == 0 );
                    var y = -( i * deltaY );
                    var leftX = -width / 2;
                    var rightX = leftX + ( isMajorTick ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH );
                    var tickNode = new Easel.Shape();
                    tickNode.graphics
                            .setStrokeStyle( 2, 'butt', 'bevel' )
                            .beginStroke( 'black' )
                            .moveTo( leftX, y )
                            .lineTo( rightX, y );
                    ticksParent.addChild( tickNode );

                    if ( isMajorTick ) {
                        // major tick label
                        var labelIndex = ( i / MINOR_TICKS_PER_MAJOR_TICK ) - 1;
                        if ( labelIndex < MAJOR_TICK_LABELS.length ) {
                            var label = StringUtils.messageFormat( Strings.pattern_0value_1units, [MAJOR_TICK_LABELS[labelIndex], Strings.units_liters] );
                            var textNode = new Easel.Text( label, "24px Arial", 'black' );
                            ticksParent.addChild( textNode );
                            textNode.x = rightX + TICK_LABEL_X_SPACING;
                            textNode.y = y - 10; //TODO replace 10 with 0.5*textNode.height
                        }
                    }
                }

                var location = mvt.modelToView( beaker.location );
                this.x = location.x;
                this.y = location.y;
            }

            // prototype chaining
            Inheritance.inheritPrototype( BeakerNode, Easel.Container );

            return BeakerNode;
        } );
